nodebuilder/share/share.go

```diff
@@ -0,0 +1,65 @@
+package share
+
+import (
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-datastore"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/routing"
+	routingdisc "github.com/libp2p/go-libp2p/p2p/discovery/routing"
+	"go.uber.org/fx"
+
+	"github.com/celestiaorg/celestia-node/service/share"
+)
+
+// LightAvailability constructs light share availability.
+func LightAvailability(cfg Config) func(
+	bServ blockservice.BlockService,
+	r routing.ContentRouting,
+	h host.Host,
+) *share.LightAvailability {
+	return func(
+		bServ blockservice.BlockService,
+		r routing.ContentRouting,
+		h host.Host,
+	) *share.LightAvailability {
+		disc := share.NewDiscovery(
+			h,
+			routingdisc.NewRoutingDiscovery(r),
+			cfg.PeersLimit,
+			cfg.DiscoveryInterval,
+			cfg.AdvertiseInterval,
+		)
+		return share.NewLightAvailability(bServ, disc)
+	}
+}
+
+// FullAvailability constructs full share availability.
+func FullAvailability(cfg Config) func(
+	bServ blockservice.BlockService,
+	r routing.ContentRouting,
+	h host.Host,
+) *share.FullAvailability {
+	return func(
+		bServ blockservice.BlockService,
+		r routing.ContentRouting,
+		h host.Host,
+	) *share.FullAvailability {
+		disc := share.NewDiscovery(
```

Comment:
could discovery be provided separately?

Reply:
Yes. To do this I had to export `share.discovery` to `share.Discovery` though, so that we can have a constructor in `share/share.go`. The reason for this is that if you change the original constructor to take a nodebuilder share config (which is needed to provide it with fx), there is a cyclical dep.